File: Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs
using System.Collections.Generic;
using System.Linq;

namespace Microsoft.Recognizers.Text.Number.Models
{
    public abstract class AbstractNumberModel : IModel
    {
        public AbstractNumberModel(IParser parser, IExtractor extractor)
        {
            this.Parser = parser;
            this.Extractor = extractor;
        }

        public abstract string ModelTypeName { get; }

        protected IExtractor Extractor { get; private set; }
        protected IParser Parser { get; private set; }

        public List<ModelResult> Parse(string query)
        {
            var extractResults = Extractor.Extract(query);
            var parseNums = new List<ParseResult>();
            foreach (var result in extractResults)
            {
                parseNums.Add(Parser.Parse(result));
            }
            return parseNums.Select(o => new ModelResult
            {
                Start = o.Start.Value,
                End = o.Start.Value + o.Length.Value - 1,
                Resolution = new SortedDictionary<string, object> { { "value", o.ResolutionStr } },
                Text = o.Text,
                TypeName = ModelTypeName
            }).ToList();
        }
    }
}

File: Java/libraries/recognizers-text-number/src/main/java/com/microsoft/recognizers/text/number/models/AbstractNumberModel.java
package com.microsoft.recognizers.text.number.models;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.IExtractor;
import com.microsoft.recognizers.text.IModel;
import com.microsoft.recognizers.text.IParser;
import com.microsoft.recognizers.text.ModelResult;
import com.microsoft.recognizers.text.ParseResult;
import com.microsoft.recognizers.text.ResolutionKey;
import com.microsoft.recognizers.text.utilities.QueryProcessor;

import java.util.ArrayList;
import java.util.List;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.stream.Collectors;

public abstract class AbstractNumberModel implements IModel {

    protected final IParser parser;
    protected final IExtractor extractor;

    protected AbstractNumberModel(IParser parser, IExtractor extractor) {
        this.parser = parser;
        this.extractor = extractor;
    }

    @Override
    public List<ModelResult> parse(String query) {

        // Pre-process the query
        query = QueryProcessor.preprocess(query, true);

        List<ParseResult> parsedNumbers = new ArrayList<ParseResult>();

        try {
            List<ExtractResult> extractResults = extractor.extract(query);
            for (ExtractResult result : extractResults) {
                ParseResult parsedResult = parser.parse(result);
                if (parsedResult != null) {
                    parsedNumbers.add(parsedResult);
                }
            }
        } catch (Exception ex) {
            // Nothing to do. Exceptions in parse should not break users of recognizers.
            // No result.
            ex.printStackTrace();
        }

        return parsedNumbers.stream().map(o -> {
            SortedMap<String, Object> sortedMap = new TreeMap<String, Object>();
            sortedMap.put(ResolutionKey.Value, o.getResolutionStr());

            // We decreased the end property by 1 in order to keep parity with other platforms (C#/JS).
            return new ModelResult(
                    o.getText(),
                    o.getStart(),
                o.getStart() + o.getLength() - 1,
                getModelTypeName(),
                sortedMap
            );                
        }).collect(Collectors.toCollection(ArrayList::new));
    }
}
